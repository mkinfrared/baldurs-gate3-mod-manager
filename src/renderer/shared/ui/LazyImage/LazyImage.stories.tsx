import { Meta, StoryObj } from "@storybook/react";

import avif1x from "../../assets/images/man-bear-pig-avif-1x.avif";
import avif2x from "../../assets/images/man-bear-pig-avif-2x.avif";
import avif3x from "../../assets/images/man-bear-pig-avif-3x.avif";
import jpg1x from "../../assets/images/man-bear-pig-jpg-1x.jpg";
import webp1x from "../../assets/images/man-bear-pig-webp-1x.webp";
import webp2x from "../../assets/images/man-bear-pig-webp-2x.webp";
import webp3x from "../../assets/images/man-bear-pig-webp-3x.webp";
import Skeleton from "../../assets/spinners/skeleton.svg";

import { LazyImage } from "./LazyImage";
import css from "./Story.module.scss";

type Story = StoryObj<typeof meta>;

const meta = {
  title: "UI/LazyImage",
  component: LazyImage,
  tags: ["autodocs"],
} satisfies Meta<typeof LazyImage>;

const Default: Story = {
  args: {
    src: "https://picsum.photos/150/200",
    srcSet:
      "https://picsum.photos/id/249/150/200 1x, https://picsum.photos/id/239/150/200 2x, https://picsum.photos/id/237/150/200 3x",
  },
  render: (args) => (
    <div className={css.Story}>
      <div className={css.def}>
        <LazyImage {...args} />
      </div>
    </div>
  ),
};

const Multiple: Story = {
  render: () => (
    <div className={css.Story}>
      <div className={css.grid}>
        <LazyImage src="https://picsum.photos/id/237/300/200" />
        <LazyImage src="https://picsum.photos/id/238/300/200" />
        <LazyImage src="https://picsum.photos/id/239/300/200" />
        <LazyImage src="https://picsum.photos/id/240/300/200" />
        <LazyImage src="https://picsum.photos/id/241/300/200" />
        <LazyImage src="https://picsum.photos/id/242/300/200" />
        <LazyImage src="https://picsum.photos/id/243/300/200" />
        <LazyImage src="https://picsum.photos/id/244/300/200" />
        <LazyImage src="https://picsum.photos/id/245/300/200" />
        <LazyImage src="https://picsum.photos/id/246/300/200" />
        <LazyImage src="https://picsum.photos/id/247/300/200" />
        <LazyImage src="https://picsum.photos/id/248/300/200" />
        <LazyImage src="https://picsum.photos/id/249/300/200" />
      </div>
    </div>
  ),
};

const Source: Story = {
  render: () => (
    <div className={css.Story}>
      <div className={css.grid}>
        <LazyImage src={jpg1x}>
          <source
            srcSet={`${avif3x} 3x, ${avif2x} 2x, ${avif1x} 1x`}
            type="image/avif"
          />
          <source
            srcSet={`${webp3x} 3x, ${webp2x} 2x, ${webp1x} 1x`}
            type="image/webp"
          />
        </LazyImage>
      </div>
    </div>
  ),
};

const SkeletonImageLoader: Story = {
  render: () => (
    <div className={css.Story}>
      <Skeleton className={css.skeleton} />
    </div>
  ),
};

export { Default, Multiple, SkeletonImageLoader, Source };

export default meta;
