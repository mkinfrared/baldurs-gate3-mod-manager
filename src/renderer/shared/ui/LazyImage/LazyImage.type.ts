import { ImgHTMLAttributes, ReactNode, RefObject } from "react";

export type LazyImageProps = ImgHTMLAttributes<HTMLImageElement> & {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  imageRef?: RefObject<HTMLImageElement>;
  children?: ReactNode;
};
