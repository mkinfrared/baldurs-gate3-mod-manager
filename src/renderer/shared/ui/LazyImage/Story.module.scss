.Story {
  .def {
    height: 200px;
    width: 150px;
  }

  .grid {
    display: grid;
    gap: 2rem;
    grid-auto-rows: 200px;
    grid-template-columns: repeat(2, 300px);
  }

  .skeleton {
    --stop-hue: 53;
    --sk-color-1: var(--primary-4);
    --sk-color-2: var(--primary-5);
  }
}
