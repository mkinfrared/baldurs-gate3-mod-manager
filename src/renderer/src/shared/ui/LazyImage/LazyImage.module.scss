.LazyImage {
  height: 100%;
  overflow: hidden;
  position: relative;
  width: 100%;

  img {
    height: 0;
    object-fit: contain;
    opacity: 0;
    width: 0;

    &.loaded {
      height: 100%;
      opacity: 1;
      transition: opacity 150ms ease-out;
      width: 100%;
    }

    &.error {
      display: none;
    }
  }

  .skeleton {
    --stop-hue: 320;
    --sk-color-1: var(--primary-4);
    --sk-color-2: var(--primary-5);

    height: 100%;
    inset: 0;

    // TODO GET RID OF POSITION ABSOLUTE
    // IT CAUSES ANIMATION ISSUES IN SAFARI ü§∑‚Äç
    // position: absolute;
    width: 100%;
  }
}
