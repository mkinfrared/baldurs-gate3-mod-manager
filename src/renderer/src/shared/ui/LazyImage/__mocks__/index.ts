export { default } from "./LazyImage.mock";
