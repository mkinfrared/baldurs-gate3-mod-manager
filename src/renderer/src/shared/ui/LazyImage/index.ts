export { default } from "./LazyImage";

export * from "./LazyImage.type";
