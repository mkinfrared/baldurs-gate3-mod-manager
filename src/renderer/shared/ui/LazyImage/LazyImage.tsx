import { memo, useRef, useState } from "react";
import { useIntersection } from "react-use";

import Skeleton from "renderer/shared/assets/spinners/skeleton.svg";
import { classNames } from "renderer/shared/lib/helpers";

import NoImage from "../NoImage";

import css from "./LazyImage.module.scss";
import { LazyImageProps } from "./LazyImage.type";

const LazyImage = ({
  alt,
  className,
  children,
  src,
  srcSet,
}: LazyImageProps) => {
  const [isLoaded, setIsLoaded] = useState(false);
  const [error, setError] = useState(false);
  const pictureRef = useRef<HTMLPictureElement>(null);
  const imageRef = useRef<HTMLImageElement>(null);

  const handleLoad = () => {
    setIsLoaded(true);
  };

  const handleError = () => {
    setIsLoaded(true);

    setError(true);
  };

  const loadImage = () => {
    const picture = pictureRef.current;
    const image = imageRef.current;

    if (!picture || !image) {
      return;
    }

    if (srcSet) {
      image.srcset = srcSet;

      return;
    }

    image.src = src ?? "";
  };

  const intersection = useIntersection(pictureRef, {
    root: null,
    rootMargin: "0px",
    threshold: 0,
  });

  if (intersection?.isIntersecting && !isLoaded) {
    loadImage();
  }

  const showChildren = isLoaded || intersection?.isIntersecting;

  return (
    <picture
      className={classNames(css.LazyImage, className)}
      ref={pictureRef}
      data-testid="LazyImage"
      onLoad={handleLoad}
      onError={handleError}
    >
      {showChildren && children}
      <img
        className={classNames(isLoaded && css.loaded, error && css.error)}
        src={undefined}
        ref={imageRef}
        alt={alt}
      />
      {!isLoaded && <Skeleton className={css.skeleton} />}
      {error && <NoImage />}
    </picture>
  );
};

export { LazyImage };

export default memo(LazyImage);
