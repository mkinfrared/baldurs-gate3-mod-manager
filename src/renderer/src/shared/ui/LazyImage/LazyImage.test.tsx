import { act, render } from "@testing-library/react";
import * as reactUse from "react-use";
import { Mocked } from "vitest";

import { LazyImage } from "./LazyImage";

vi.mock("react-use");

describe("<LazyImage />", () => {
  const hooksMock = reactUse as Mocked<typeof reactUse>;
  const src = "marklar";
  const children = <source data-testid="source" />;
  const Component = <LazyImage src={src}>{children}</LazyImage>;

  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("should be defined", () => {
    expect(LazyImage).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("LazyImage");

    expect(element).toBeDefined();
  });

  it("should render an image without 'src' attribute and not render children", () => {
    const { container, queryByTestId } = render(Component);
    const image = container.querySelector("img") as HTMLImageElement;
    const source = queryByTestId("source");

    expect(image.src).toBe("");

    expect(source).toBe(null);
  });

  it("should render children when 'isIntersecting' is true", async () => {
    hooksMock.useIntersection.mockReturnValue({
      isIntersecting: true,
    } as any);

    const { findByTestId, rerender } = render(Component);

    rerender(Component);

    const source = await findByTestId("source");

    expect(source).toBeDefined();
  });

  it("should render <NoImage /> when image load fails", async () => {
    hooksMock.useIntersection.mockReturnValue({
      isIntersecting: true,
    } as any);

    const event = new Event("error", { bubbles: true });

    const { container, findByTestId, rerender } = render(
      <LazyImage src={src}>{children}</LazyImage>,
    );

    const image = container.querySelector("img");

    rerender(<LazyImage src={src}>{children}</LazyImage>);

    act(() => {
      image?.dispatchEvent(event);
    });

    const noImage = await findByTestId("NoImage");

    expect(noImage).toBeDefined();
  });
});
