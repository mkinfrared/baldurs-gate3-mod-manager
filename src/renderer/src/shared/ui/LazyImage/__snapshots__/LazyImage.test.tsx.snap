// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LazyImage /> > should match the snapshot 1`] = `
<div>
  <picture
    class="LazyImage"
    data-testid="LazyImage"
  >
    <img
      class=""
    />
    <svg
      class="skeleton"
    >
      skeleton.svg
    </svg>
  </picture>
</div>
`;
